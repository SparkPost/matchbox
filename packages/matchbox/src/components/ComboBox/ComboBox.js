import React from 'react';
import PropTypes from 'prop-types';

function ComboBox(props) {
  const { children, style, rootRef, ...rest } = props;

  return (
    <div ref={rootRef} style={{ position: 'relative', ...style }} {...rest}>
      {children}
    </div>
  );
}

ComboBox.propTypes = {
  children: PropTypes.node,
  style: PropTypes.object,
  /**
   * Maps to Downshift's refKey set in getRootProps. refKey must be set to "rootRef"
   */
  rootRef: PropTypes.func
};

export default ComboBox;
