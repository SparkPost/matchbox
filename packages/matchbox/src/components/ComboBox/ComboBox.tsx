import React, { ReactNode } from 'react';
import { getChild } from '../../helpers/children';
import { Box } from '../Box';

export type ComboBoxProps = {
  children?: ReactNode;
  style?: React.CSSProperties;
  /**
   * Maps to Downshift's refKey set in getRootProps. refKey must be set to "rootRef"
   */
  rootRef?: React.MutableRefObject<HTMLDivElement>;
};

function ComboBox(props: ComboBoxProps): JSX.Element {
  const { children, style, rootRef, ...rest } = props;

  // Clones children in order to pass the menu into the textfield
  // Textfield inserts the menu in the DOM directly after the input, before helptext and error
  const textfieldWithMenu = React.useMemo(() => {
    return getChild('ComboBoxTextField', { children: getChild('ComboBoxMenu', {}) });
  }, [children]);

  return (
    <Box position="relative" ref={rootRef} style={style} {...rest}>
      {textfieldWithMenu}
    </Box>
  );
}

export default ComboBox;
