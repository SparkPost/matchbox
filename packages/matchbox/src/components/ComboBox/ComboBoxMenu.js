import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { ActionList } from '../ActionList';
import styles from './ComboBoxMenu.module.scss';

function ComboBoxMenu(props) {
  const { items, menuRef, isOpen, maxHeight, ...rest } = props;

  const listClasses = classnames(
    styles.List,
    isOpen && styles.open
  );

  return (
    <div {...rest} ref={menuRef} className={listClasses}>
      <ActionList actions={items} maxHeight={maxHeight} />
    </div>
  );
}

ComboBoxMenu.propTypes = {
  /**
   * Array of items. see ActionList prop types.
   */
  items: PropTypes.array,
  /**
   * Controls menu visibility
   */
  isOpen: PropTypes.bool,
  /**
   * Maps to Downshift's refKey set in getMenuProps. refKey must be set to "menuRef"
   */
  menuRef: PropTypes.func
};

ComboBoxMenu.defaultProps = {
  items: []
};

export default ComboBoxMenu;
