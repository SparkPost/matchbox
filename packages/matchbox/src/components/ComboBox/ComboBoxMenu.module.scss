@import '../../styles/config';

.List {
  position: absolute;
  left: 0;
  right: 0;
  top: 100%;
  z-index: z-index(overlay);

  background: color(gray, 10);
  border-radius: border-radius();
  border: 1px solid color(gray, 7);
  box-shadow: shadow(deep);

  margin-top: 3px;
  max-height: 20rem;

  overflow-y: scroll;
  pointer-events: none;

  transform: scaleY(0.97) translate(0, -5px);
  opacity: 0;
  transition: 0s ease-out;

  &.open {
    opacity: 1;
    pointer-events: auto;
    transform: scaleY(1) translate(0, 0);
    transition: 0.2s ease-out;
  }
}
