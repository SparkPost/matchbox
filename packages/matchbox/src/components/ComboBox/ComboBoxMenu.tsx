import React from 'react';
import { MaxHeightProps } from 'styled-system';
import { ActionList, ActionListActionProps } from '../ActionList';
import { Box } from '../Box';
import PopoverContent from '../Popover/PopoverContent';

export type ComboBoxMenuProps = {
  /**
   * String displayed when item length is 0. Defaults to "No Results"
   */
  emptyMessage?: string;
  /**
   * Array of items. see ActionList prop types.
   */
  items?: ActionListActionProps[];
  /**
   * Controls menu visibility
   */
  isOpen?: boolean;
  /**
   * Maps to Downshift's refKey set in getMenuProps. refKey must be set to "menuRef"
   */
  menuRef?: React.MutableRefObject<HTMLDivElement>;
} & MaxHeightProps;

function ComboBoxMenu(props: ComboBoxMenuProps): JSX.Element {
  const {
    items = [],
    menuRef,
    isOpen,
    maxHeight = '20rem',
    emptyMessage = 'No Results',
    ...rest
  } = props;

  return (
    <Box ref={menuRef} {...rest} position="relative" zIndex="overlay">
      <PopoverContent open={isOpen} width="100%">
        {items.length > 0 ? (
          <ActionList maxHeight={maxHeight}>
            {items.map((action, i) => (
              <ActionList.Action key={i} {...action} />
            ))}
          </ActionList>
        ) : (
          <Box p="200" color="gray.700">
            {emptyMessage}
          </Box>
        )}
      </PopoverContent>
    </Box>
  );
}

export default ComboBoxMenu;
