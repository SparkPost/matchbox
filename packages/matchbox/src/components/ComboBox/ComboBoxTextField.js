import React from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import { Error } from '../Error';
import { Tag } from '../Tag';
import { identity, noop } from '../../helpers/event';
import { onKey } from '../../helpers/keyEvents';
import classnames from 'classnames';
import styles from './ComboBoxTextField.module.scss';

function ComboBoxTextField(props) {
  const {
    autoFocus,
    disabled,
    error,
    errorInLabel,
    helpText,
    id,
    itemToString,
    label,
    labelHidden,
    name,
    onChange,
    onFocus,
    onBlur,
    onKeyDown,
    placeholder,
    readOnly,
    removeItem,
    style,
    selectedItems,
    value,
    ...rest
  } = props;

  const inputRef = React.useRef();

  const setWrapperClasses = classnames(
    styles.Wrapper,
    error && styles.Error,
    disabled && styles.Disabled
  );

  const labelMarkup = (
    <Label id={id} label={label}>
      {error && errorInLabel && <Error className={styles.InlineError} wrapper='span' error={error} />}
    </Label>
  );

  const helpMarkup = helpText
    ? <div className={styles.HelpText}>{helpText}</div>
    : null;

  const selectedMarkup = selectedItems.length
    ? selectedItems.map((item, i) => (
      <Tag
        key={i}
        onRemove={!disabled
          ? () => removeItem(item)
          : null
        }
      >
        {itemToString(item)}
      </Tag>
    ))
    : null;

  // Auto focuses the input
  function handleClick() {
    inputRef.current.focus();
  }

  // Removes last item with a backspace and and empty input value
  function handleKeyDown(e) {
    if (!value && selectedItems.length) {
      onKey('backspace', () => {
        removeItem(selectedItems[selectedItems.length - 1]);
      })(e);
    }

    if (onKeyDown) {
      onKeyDown(e);
    }
  }

  return (
    <fieldset className={styles.TextField}>
      {label && !labelHidden && labelMarkup}
      <div className={setWrapperClasses} onClick={handleClick}>
        {selectedMarkup}
        <input
          {...rest}
          autoFocus={autoFocus}
          className={styles.Input}
          disabled={disabled}
          id={id}
          name={name}
          onFocus={onFocus}
          onBlur={onBlur}
          onChange={onChange}
          onKeyDown={handleKeyDown}
          placeholder={placeholder}
          readOnly={readOnly}
          ref={inputRef}
          style={style}
          value={value}
        />
      </div>
      {error && !errorInLabel && <Error error={error} />}
      {helpMarkup}
    </fieldset>
  );
}

ComboBoxTextField.propTypes = {
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  errorInLabel: PropTypes.bool,
  helpText: PropTypes.node,
  id: PropTypes.string,
  itemToString: PropTypes.func,
  label: PropTypes.string,
  labelHidden: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  removeItem: PropTypes.func,
  selectedItems: PropTypes.array
};

ComboBoxTextField.defaultProps = {
  selectedItems: [],
  itemToString: identity,
  removeItem: noop
};

export default ComboBoxTextField;
