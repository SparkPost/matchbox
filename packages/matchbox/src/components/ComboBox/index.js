export { default as ComboBox } from './ComboBox';
export { default as ComboBoxMenu } from './ComboBoxMenu';
export { default as ComboBoxTextField } from './ComboBoxTextField';
