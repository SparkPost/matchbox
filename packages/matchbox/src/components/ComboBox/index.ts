export { default as ComboBox } from './ComboBox';
export { default as ComboBoxMenu } from './ComboBoxMenu';
export { default as ComboBoxTextField } from './ComboBoxTextField';
export * from './ComboBox';
export * from './ComboBoxMenu';
export * from './ComboBoxTextField';
