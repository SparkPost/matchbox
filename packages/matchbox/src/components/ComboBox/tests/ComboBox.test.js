import React from 'react';
import ComboBox from '../ComboBox';
import { mount } from 'enzyme';

describe('ComboBox Wrapper', () => {
  const subject = (props = {}) => mount(<ComboBox {...props}/>);

  it('should render styles, children, and other props correctly', () => {
    expect(subject({
      children: <span>test children</span>,
      style: { background: 'red' },
      onClick: jest.fn()
    }).find('div')).toMatchSnapshot();
  });
});
