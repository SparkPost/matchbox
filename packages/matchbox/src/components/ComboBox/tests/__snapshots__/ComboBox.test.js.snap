// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox Wrapper should render styles, children, and other props correctly 1`] = `
<div
  onClick={[MockFunction]}
  style={
    Object {
      "background": "red",
      "position": "relative",
    }
  }
>
  <span>
    test children
  </span>
</div>
`;
